Additional Configurability of FlatFileStorage

Wealthyturtle opened the pull request:
### Description
Allows users to configure the variable change threshold to trigger a re-save of the database file, as well as how often the database file is re-written.

---
**Target Minecraft Versions:**     Any
**Requirements:**     None
**Related Issues:**     #2007

bensku commented on src/main/java/ch/njol/skript/variables/FlatFileStorage.java:
```diff
@@ -191,9 +194,38 @@ protected boolean load_i(final SectionNode n) {
 			Skript.info(file.getName() + " successfully updated.");
 		}
 		
+		String thresholdRaw = getValue(n, "variable re-save threshold");
```
Use getValue overload that takes in type.

bensku commented on src/main/java/ch/njol/skript/variables/FlatFileStorage.java:
```diff
@@ -191,9 +194,38 @@ protected boolean load_i(final SectionNode n) {
+			}
+		}
+		
+		String rewriteRaw = getValue(n, "file re-write frequency in ticks");
```
Same here.

bensku commented on src/main/resources/config.sk:
```diff
@@ -265,6 +265,15 @@ databases:
+		
+		file re-write frequency in ticks: 6000
+		# The number of ticks that should pass before the database file is re-written.
+		# Setting this value lower is not recommended as saving large database files may cause server lag.
```
Comment should explicitly tell how low it is safe to set. Also, I don't think the description about lag is quite right. Saving large databases takes *time*, and bad things happen if a database save has not succeeded before next one starts.

Blueyescat replied:
>  bad things happen if a database save has not succeeded before next one starts.

that should be prevented

bensku commented on src/main/java/ch/njol/skript/variables/FlatFileStorage.java:
```diff
@@ -99,6 +105,7 @@ protected boolean load_i(final SectionNode n) {
 		boolean update2_0_beta3 = false;
 		final Version v2_1 = new Version(2, 1);
 		boolean update2_1 = false;
+		final Version v_2_4_beta9 = new Version(2, 4, "beta 9");
```
This PR is staged for Skript 2.5. Just check for 2.4 or older version.

Blueyescat commented on src/main/java/ch/njol/skript/variables/FlatFileStorage.java:
```diff
@@ -84,7 +90,7 @@ protected FlatFileStorage(final String name) {
 	/**
 	 * Doesn'ts lock the connection as required by {@link Variables#variableLoaded(String, Object, VariablesStorage)}.
```
Doesn't

Blueyescat commented on src/main/java/ch/njol/skript/variables/FlatFileStorage.java:
```diff
@@ -191,9 +198,62 @@ protected boolean load_i(final SectionNode n) {
+					convertedCfg.add(line);
+					if (line.contains("type: CSV") || line.contains("type: SQLite")) {
+						convertedCfg.add("");
+						convertedCfg.add("		variable re-save threshold: 1000");
```
This option should come after the other one, :)

Blueyescat commented:
I think the option names aren't user friendly.
```diff
- variable re-save threshold
+ required variable changes for save
```

and this should work with timespans
```diff
- file re-write frequency in ticks
+ variable save interval
```

Blueyescat commented on src/main/java/ch/njol/skript/variables/FlatFileStorage.java:
```diff
@@ -191,9 +198,62 @@ protected boolean load_i(final SectionNode n) {
+				String line = null;
+				while((line = r.readLine()) != null) {
+					convertedCfg.add(line);
+					if (line.contains("type: CSV") || line.contains("type: SQLite")) {
```
Bad check

Blueyescat replied:
Wait, you are adding these options into the default storage setting section, but the saving code is global.
Don't forget that there can be multiple flat file storages (default is one of them with name 'default'), and a FlatFileStorage is equal to a single storage.

I mean, the saving code should be specific to a single storage (see how other storage specific options work), or at least the options should be in the global section.

Wealthyturtle replied:
Does the current code not work with that? Isn't a SectionNode referring to a section of the config and not the whole thing?

Wealthyturtle replied:
Just tested, this code supports multiple flat file databases, with each database having its own save interval and min. variable threshold. However, I do agree that the check has issues and needs to be reworked.

Wealthyturtle commented:
Current Changes:
- Updated documentation of SkriptConfig.getPreviousConfigVersion()
- Added nullness to SkriptConfig.getPreviousConfigVersion(), in the case that for whatever reason the previous config version could not be derived
- Fixed typo in documentation of FlatFileStorage.load_i()
- Replaced null suppression with resource suppression in FlatFileStorage.load_i()
- Added resource suppression in FlatFileStorage.saveVariables() to make the file warning-less
- Updated previous config version testing to before version 2.5
- Renamed new options 'variable re-save threshold' and 'file re-write frequency in ticks' to 'required variable changes for save' and 'variable save interval' respectively
- Rewrote description of the 'required variable changes for save' and 'variable save interval' options to mention flat file databases
- Added a disclaimer for the 'variable save interval' option that setting a value to 1200 ticks or lower is not advisable as it may cause bad things to happen if a database save has not succeeded before next one starts

bluelhf commented:
@Wealthyturtle Any updates on this?

FranKusmiruk commented:
damn, tried to resolve conflicts and somehow this happened, going to try to fix it

FranKusmiruk commented:
Closing in favour of #3786
